Labs 2-1 and 2-2 for Building and Executing content on AAP (#8)
rhpds/ansible-super-lab-showroom: documentation/modules/ROOT/pages/21-content-dev-ee.adoc

. Launch the link:{aap_controller_web_url}[Private Automation Hub,window=_blank] web interface and login using the credentials from the xref:environment-details.adoc[Environment Details,window=_blank] page.
. From the navigation menu on the left, expand **Automation Content** and click on **API Token**.
+
image::21-content-dev-ee/api-token-overview.png[API Token Overview]
+
. On the API Token page, click the **Generate token** button. Your API token will be displayed.
+
image::21-content-dev-ee/api-token-generate.png[Generate an API Token]
+
. Click the *Copy to clipboard* icon to copy it.
+
NOTE: Make sure to store the token securely, as it will not be displayed again.

==== Step 2.1.3: Set Environment Variables

In your Dev Spaces terminal, create a file containing the following variables. These will configure `ansible-galaxy` and `ansible-builder` to use your PAH instance instead of upstream to pull and push collections and execution environments.

> . First, create a file in your `/projects/env/` directory with the following contents, be sure to use your PAH API token created previously:

Reviewer: ```suggestion
. First, create a new directory named `env` in the `/projects` directory. 

[source,bash,role=execute,subs="verbatim,attributes,quotes"]
----
mkdir /projects/env
----

. Now, create a file called `set_pah_vars.env` in your `/projects/env/` directory with the following contents. Be sure to use your PAH API token created previously:
```

Double check the numbering after the proposed changes above

. Launch the link:{aap_controller_web_url}[Private Automation Hub,window=_blank] web interface and login using the credentials from the xref:environment-details.adoc[Environment Details,window=_blank] page.

> . From the navigation menu on the left, expand **Automation Content** and click on **Namespaces**.

Reviewer: Not specifically related to this line, but the section title for 3.2.1 sounds a bit weird. Thoughts on updating it to "3.2.1: Create a namespace for the collection"?

. From the navigation menu on the left, expand **Automation Content** and click on **Collection Approvals**.
+
image::21-content-dev-ee/collections-approvals-pending.png[Approve the Collection]
+
. The `red_hat_one.super_lab` should be listed as pending approval. Select this namespace checkbox and click the **Thumbs Up** icon all the way to the right to approve and sign the collection.
+
image::21-content-dev-ee/collections-approvals-selection.png[Select Collection to Approve]
+
. Select the _Yes, I confirm that I want to approve these 1 collections._ checkbox and click **Approve collections** button.
+
image::21-content-dev-ee/collections-approvals-approve-collection.png[Pending Approval Collection]
+
. Navigate to the **Collections** page in PAH to verify that your collection is published and available.
+
. You should see your `red_hat_one.super_lab` collection listed and available for use.
+
image::21-content-dev-ee/pah-collections.png[PAH Collections]

== 4: Syncing a Base EE from the Red Hat Registry

> Before producing our own Execution Environment (EE) in subsequent steps, we'll configure PAH to pull in a certified base image from Red Hat. Since AAP images within the Red Hat Container Registry are private, we need to configure a credential within PATH to allow for the retrieval of images.

Reviewer: ```suggestion
Before producing our own Execution Environment (EE) in subsequent steps, we'll configure PAH to pull in a certified base image from Red Hat. Since AAP images within the Red Hat Container Registry are private, we need to configure a credential within PAH to allow for the retrieval of images.
```

+
image::21-content-dev-ee/create-execution-environment-form.png[Create Execution Environments Form]
+
. Synchronize the Execution Environment into PAH from the Red Hat Container Registry.
. Next to the `ansible-automation-platform-26/ee-minimal-rhel9` Execution Environment, click the kebab menu (three vertical dots) on the right hand side of the entry and select **Sync execution environment**.
+
image::21-content-dev-ee/execution-environment-sync.png[Sync Execution Environment]
+
. Click the checkbox next to the _Yes, I confirm I want to sync these 1 execution environments_.  the box to confirm and click **Sync execution environment**.
. Click the **Sync Execution Environments** button to start the synchronization process.
+
image::21-content-dev-ee/execution-environment-confirm-sync.png[Confirm Sync Execution Environment]
+
. After the sync is complete, the `ee-minimal-rhel9` image will be available in your Private Automation Hub instance. Feel free to explore the details of the newly synchronized Execution Environment as you see fit.

=== 4.3: Synchronize required collections

. Go to *Automation Content* -> *Remotes*

> . Click on the kebob (3 dots) to the right of *rh-certified* and choose the edit (pencil) icon

Reviewer: ```suggestion
. Click on the edit (pencil) icon to the right of *rh-certified* remote
```

+
image::21-content-dev-ee/edit-certified-remote.png[Edit RH-Certified Repository]
+
. In the *Token* field, paste the token obtained in the previous lab for upstream Automation Hub
. In the *Requirements file* field, paste the contents of `/projects/lab_playbooks/collections/requirements.yml`
+
[source,yaml]
----
collections:
  - chocolatey.chocolatey
  - ansible.windows
  - microsoft.iis
----
+
. Click *Save remote*
+
image::21-content-dev-ee/edit-certified-details.png[Edit Details]
+

> . Navigate to *Automation Content* -> *Repositories*, next to *rh-certified* click the kebob (3 dots) and select *Sync repository*

Reviewer: Theres a dialog asking whether to _Mirror_ and/or _Optimize_ the Sync. We should include steps on what actions to take

. Navigate to your link:{gitea_console_url}[Gitea instance,window=_blank] and click the **Sign In** button on the upper right hand corner
+ 
image::21-content-dev-ee/gitea-homepage.png[Gitea Homepages]
+
. Enter the username and password using the credentials provided from the xref:environment-details.adoc[Environment Details,window=_blank] page and click the **Sign In** button
+ 
image::21-content-dev-ee/gitea-login.png[Gitea Login Page]
+
. Once authenticated, in the top left of the web interface, click on the **+** symbol and select **New Repository**.
+
image::21-content-dev-ee/gitea-navigate-new-repository.png[Navigate to new Gitea Repository]
+

> . On the New Repository page, enter 'red_hat_one_super_lab' in the Repository Name field.

Reviewer: ```suggestion
. On the New Repository page, enter `red_hat_one_super_lab` in the Repository Name field.
```

+
[source,bash,role=execute,subs="verbatim,attributes"]
----
podman push aap-aap.{openshift_cluster_ingress_domain}/my-pah-ee:1.1
----
+
[source,output]
----
Getting image source signatures
Copying blob e9b18b42db30 done   | 
Copying blob 3e44b8c037a4 done   | 
Copying blob c644513b7c21 done   | 
Copying blob 03c21ed8e101 done   | 
Copying blob eaaedab2167d done   | 
Copying blob e189a7d08ffb done   | 
Copying blob aa5eefe60148 skipped: already exists  
Copying blob 4b8fa9ef238e done   | 
Copying blob 25e49599071c done   | 
Copying blob 5f70bf18a086 done   | 
Copying blob 9f1b4907d130 done   | 
Copying blob b026c973af3b done   | 
Copying blob 0f9da8a96a25 done   | 
Copying config 81a91a3d03 done   | 
----
+
. Verify by navigating in AAP controller to *Automation Content* -> *Execution Environments* -> *my-pah-ee* -> *Images*
+
image::21-content-dev-ee/execution-environment-versions.png[View EE Versions]

== 8: Test content in Dev Spaces

Now that both the collection and Execution Environment have been created and published to Private Automation hub, we'll create a playbook to use the filter contained within the `red_hat_one.super_lab` collection.

=== 8.1: Create a Playbook

> Since we already have a plabyook directory at `/projects/lab_playbooks/` we will use this.

Reviewer: ```suggestion
Since we already have a playbook directory at `/projects/lab_playbooks/` we will use this location.
```

  tasks:
    - name: Print a message using the cowsay filter
      ansible.builtin.debug:
        msg: "{{ 'Hello from my custom filter!' | red_hat_one.super_lab.cowsay }}"
...
----

=== 8.2: Update Collections requirements

Since this playbook requires the custom collection, let's add it to the requirements file:

[source,yaml,title="/projects/lab_playbooks/collections/requirements.yml",role=execute,subs="verbatim,attributes"]
----
collections:
  - chocolatey.chocolatey
  - ansible.windows
  - microsoft.iis
  - red_hat_one.super_lab<1>
----
<1> Add the custom collection here. Note that due to the symlink we created earlier, this isn't required, but will be when running on AAP.

=== 8.3: Run playbook

. Attempt to run this locally. It will fail. Do you know why?
+

> [source,yaml,role=execute,subs="verbatim,attributes"]

Reviewer: ```suggestion
[source,bash,role=execute,subs="verbatim,attributes"]
```

+
image::21-content-dev-ee/automation-controller-create-ee.png[Create Execution Environment]
+
. Enter the following details on the _Create Execution Environment_ page:
.. Name: `My Custom PAH EE`
.. Image URL: `aap-aap.{openshift_cluster_ingress_domain}/my-pah-ee:1.1`
.. Registry Credential: Select the `Hub` credential created previously.
. Click **Create execution environment** to create the new EE.

image::21-content-dev-ee/automation-controller-create-ee-dialog.png[Create Execution Environment Dialog]

=== 9.5: Create a Project

Create a new link:https://docs.ansible.com/ansible-tower/latest/html/userguide/projects.html[Project,window=_blank] in Automation Controller that points to your Git repository containing the playbook.

. From the navigation menu on the left, expand **Automation Execution** and click on **Projects**.
+
image::21-content-dev-ee/automation-controller-projects-overview.png[Navigate to Projects]
+
. Click **Create project** to add a new project.
+
image::21-content-dev-ee/automation-controller-create-project.png[Create Project]
+
. Enter the following details on the _Create Project_ page:

> .. Name: `Custom Collection"

Reviewer: Shouldnt this be `Lab Playbooks`?

.. Organization: `Default`
.. Source Control Type: `Git`

> .. Source control URL: `{gitea_console_url}/{gitea_user}/red_hat_one_super_lab.git`

Reviewer: ```suggestion
.. Source control URL: `{gitea_console_url}/{gitea_user}/lab_playbooks.git`
```

.. Source code credential: Select the `Gitea` credential created previously.
. Click **Create project** to create the new Project.

image::21-content-dev-ee/automation-controller-create-project-dialog.png[Create Project Dialog]

Confirm that the project has been synced successfully by looking _Success_ underneath _Last job status_.

image::21-content-dev-ee/automation-controller-sync-project.png[Project Sync Status]

=== 9.6: Create a Job Template

With the Execution Environment and Project created, the final step is to create a Job Template that uses both components.

. From the navigation menu on the left, expand **Automation Execution** and click on **Templates**.
+
image::21-content-dev-ee/automation-controller-templates-overview.png[Navigate to Templates]
+
. Click the **Create template** dropdown and select **Create job template**.
+
image::21-content-dev-ee/automation-controller-create-job-template.png[Create Job Template]
+
. Enter the following details on the _Create job template_ page:
.. Name: `Test Custom Cowsay Filter`
.. Inventory: `super-lab`
.. Project: `Lab Playbooks`

> .. Playbook: `playbooks/test_pah_ee.yml`

Reviewer: ```suggestion
.. Playbook: `test_pah_ee.yml`
```
